pydantic_ai_graph/pydantic_ai_graph/nodes.py

```diff
@@ -1,113 +1,93 @@
 from __future__ import annotations as _annotations
 
-from abc import ABC, ABCMeta, abstractmethod
+from abc import abstractmethod
 from dataclasses import dataclass
 from functools import cache
-from typing import Any, ClassVar, Generic, get_args, get_origin, get_type_hints
+from typing import Any, ClassVar, Generic, get_origin, get_type_hints
 
-from typing_extensions import TypeVar
+from typing_extensions import Never, TypeVar
 
 from . import _utils
 from .state import StateT
 
-__all__ = (
-    'NodeInputT',
-    'GraphOutputT',
-    'GraphContext',
-    'End',
-    'BaseNode',
-    'NodeDef',
-)
+__all__ = ('GraphContext', 'End', 'BaseNode', 'NodeDef')
 
-NodeInputT = TypeVar('NodeInputT', default=Any)
-GraphOutputT = TypeVar('GraphOutputT', default=Any)
+RunEndT = TypeVar('RunEndT', default=None)
+NodeRunEndT = TypeVar('NodeRunEndT', covariant=True, default=Never)
 
 
-# noinspection PyTypeHints
 @dataclass
 class GraphContext(Generic[StateT]):
     """Context for a graph."""
 
     state: StateT
 
 
-# noinspection PyTypeHints
-class End(ABC, Generic[NodeInputT]):
+@dataclass
+class End(Generic[RunEndT]):
     """Type to return from a node to signal the end of the graph."""
 
-    __slots__ = ('data',)
-
-    def __init__(self, input_data: NodeInputT) -> None:
-        self.data = input_data
+    data: RunEndT
 
-
-class _BaseNodeMeta(ABCMeta):
-    def __repr__(cls):
-        base: Any = cls.__orig_bases__[0]  # type: ignore
-        args = get_args(base)
-        if len(args) == 3 and args[2] is Any:
-            if args[1] is Any:
-                args = args[:1]
-            else:
-                args = args[:2]
-        args = ', '.join(_utils.type_arg_name(a) for a in args)
-        return f'{cls.__name__}({base.__name__}[{args}])'
+    @classmethod
+    def get_id(cls) -> str:
+        return 'End'
 
 
-# noinspection PyTypeHints
-class BaseNode(Generic[StateT, NodeInputT, GraphOutputT], metaclass=_BaseNodeMeta):
+class BaseNode(Generic[StateT, NodeRunEndT]):
     """Base class for a node."""
 
-    node_id: ClassVar[str | None] = None
-    __slots__ = ('input_data',)
-
-    def __init__(self, input_data: NodeInputT) -> None:
-        self.input_data = input_data
+    _node_id: ClassVar[str | None] = None
```

Comment:
does this have to be private? I get it can be but think it'll be clearer that it's part of the public api if it's public.

Reply:
It can be part of the public API, but the problem is that in lots of places we were accessing it directly, rather than via `get_id()` which will grab the actual string, and which needs to be used everywhere (or else you'll get `None`s and therefore not unique ids). I changed it to private to make that mistake less likely but I understand your point that now it looks like users aren't supposed to set it. Maybe we can call it `node_id_override` instead? Or we can just keep it as `node_id` but basically I just don't want to make that mistake of accessing that value instead of calling `node.get_id()`. Maybe we can set it in the metaclass if it doesn't have an explicitly-set value?

Reply:
I wonder if it wouldn't just be better to drop the attribute entirely and have people override `get_id` if they want to manually change the value. That would ensure it's not possible to use the wrong api, for the in-my-opinion small cost of requiring a little more verbosity if you want to override the value. Given I expect we'll generally discourage overriding the value that seems okay to me

Reply:
I'm imagining adding a diff that looks like this:
```diff
     @classmethod
     @cache
     def get_id(cls) -> str:
-        return cls._node_id or cls.__name__
+        """Get the ID of the node.
+
+        You can override this if you want to serialize/deserialize nodes using a specific ID.
+        """
+        return cls.__name__
 ```

Reply:
I removed `_node_id` for now, didn't add that docstring since we don't have node serialization/deserialization now and not sure that's how it will work. Can undo this change if preferred though.